module/zfs/arc.c

```diff
@@ -5193,58 +5200,65 @@ arc_get_data_buf(arc_buf_hdr_t *hdr, uint64_t size, void *tag)
 void
 arc_wait_for_eviction(uint64_t amount)
 {
-	mutex_enter(&arc_evict_lock);
-	if (arc_is_overflowing()) {
-		arc_evict_needed = B_TRUE;
-		zthr_wakeup(arc_evict_zthr);
-
-		if (amount != 0) {
-			arc_evict_waiter_t aw;
-			list_link_init(&aw.aew_node);
-			cv_init(&aw.aew_cv, NULL, CV_DEFAULT, NULL);
+	switch (arc_is_overflowing()) {
+	case 0:
+		return;
+	case 1:
+		if (!arc_evict_needed) {
+			arc_evict_needed = B_TRUE;
+			zthr_wakeup(arc_evict_zthr);
+		}
+		return;
+	default:
```

Comment:
It would add another branching and the panic code itself, while in this case it is really safe (just slower) to take the default path.

Reply:
I agree that the current code is correct.  My concern is that if someone adds a new variant of the enum that has some different meaning, we want to make sure that they think about how that should be handled here, instead of just treating it the same as ARC_OVF_SEVERE.  In other words, it's generally good practice for switch statements to handle all the variants and panic if an unknown variant is encountered - it's defensive against future changes.  But I see that there are plenty of places in the existing code that don't do that, so oh well.

Reply:
On FreeBSD there is __assert_unreachable() macro, turning into panic() when built with debug and into __builtin_unreachable() otherwise if compiler supports it.  We could introduce something like that in ZFS, if somebody knows how to properly spell it on Linux.

Reply:
I'm not sure I understand the argument against `default: panic("invalid arc_ovf_level_t")`

Reply:
It is a waste of time for additional branching and a code trashing.

Reply:
I disagree.  I think any performance cost is trivial (but will happily change my mind if you have performance measurements to the contrary).  In my opinion, indicating what states are expected and allowed makes the code more clear.

Reply:
Lets say I saw in some profiles before effects of alike additional cases.  Depending on situation compiler may turn switch into number of if's, and additional branching in a hot path never helps.  In this particular case I believe compiler will inline arc_is_overflowing(), since it is now static and used in only one place, and just throw out any additional code we'd add there.

I've described above how I see it to be done properly, both paranoid in debug builds, may be even more optimized in production and clearly visible in the code.